src/System.Text.Primitives/System/Text/Encoding/Utf8/Utf8Encoder.cs

```diff
@@ -288,55 +298,121 @@ internal static bool TryComputeEncodedBytes(ReadOnlySpan<uint> utf32, out int by
         /// <param name="charactersConsumed">On exit, contains the number of characters that were consumed from the UTF-16 span.</param>
         /// <param name="bytesWritten">An output parameter to store the number of bytes written to <paramref name="utf8"/></param>
         /// <returns>True if the input buffer was fully encoded into the output buffer, otherwise false.</returns>
-        public static bool TryEncode(ReadOnlySpan<char> utf16, Span<byte> utf8, out int charactersConsumed, out int bytesWritten)
+        public static unsafe bool TryEncode(ReadOnlySpan<char> utf16, Span<byte> utf8, out int charactersConsumed, out int bytesWritten)
         {
-            var availableCharacters = utf16.Length;
-            var bufferLength = utf8.Length;
-
-            charactersConsumed = 0;
-            bytesWritten = 0;
-
-            for (int i = 0; i < availableCharacters; i++)
+            fixed (char* pUtf16 = &utf16.DangerousGetPinnableReference())
             {
-                var ch = utf16[i];
-
-                if ((ushort)ch <= 0x7f) // this if block just optimizes for ascii
+                fixed (byte* pUtf8 = &utf8.DangerousGetPinnableReference())
                 {
-                    if (bufferLength - bytesWritten < 1)
-                        return false;
+                    var outputBytesAvailable = utf8.Length;
+                    var inputCharacters = utf16.Length;
 
-                    utf8[bytesWritten++] = (byte)ch;
-                    charactersConsumed++;
-                }
-                else
-                {
-                    uint codePoint;
-                    int consumed;
+                    var bytesWrittenTemp = 0;
+                    var i = 0;
+                    var codePoint = '\0';
 
-                    if (!char.IsSurrogate(ch))
+                    ascii:
+                    for (; i < inputCharacters; i++)
                     {
-                        codePoint = (uint)ch;
-                        consumed = 1;
+                        codePoint = *(pUtf16 + i);
+
+                        if (codePoint > Utf8OneByteLastCodePoint)
+                        {
+                            goto non_ascii;
+                        }
+                        if (bytesWrittenTemp > outputBytesAvailable - 1)
+                        {
+                            goto need_more;
+                        }
```

Comment:
Yep, that's what the CoreCLR impl does. 

Once you start doing some of these optimizations, you will likely hit problems with register pressure: things will become slower because they won't fit into registers anymore. The fix for that is to use `pointer` that fits into one register instead of `base+index` that takes two registers.